### ISSUE TITLE
PowerShell on osx how to force to be completely monochrome

### ISSUE THREAD

**Issue by rhubarb-geek-nz** [PowerShell/PowerShell#21066]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

My Terminal window is black text on white background - aka the default at this time of day.

The PowerShell prompt is printed in black but my typed text is yellow.

How do I switch PowerShell to -

- Not use any colours at all, just be monochrome
- automatically switch between black and white, or white on black along with the macOS terminal automatically changing its theme based on time of day

### Expected behavior

```console
I would like to easily see what I have typed
```

### Actual behavior

```console
Typed text is yellow on a white background which is very hard to see
```

### Error details

```console
none
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.1
PSEdition                      Core
GitCommitId                    7.4.1
OS                             Darwin 23.2.0 Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:34 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T8103
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

Yellow text on white background

**Comment by 237dmitry** [comment]
You can set all colors of `PSStyle` and `Set-PSReadline -Color` in external script files and dot-sourced them from `$profile`. For example:

```powershell
switch ([int][datetime]::Now.DayOfWeek)
{
    0 { . mon.ps1 }
    1 { . tue.ps1 }
    2 { . wed.ps1 }
...
    6 { . sun.ps1 }
}
```

**Comment by rhubarb-geek-nz** [comment]
Hm, I don't want PowerShell to do anything at all with the colours as if it was a VT102 terminal with no colours at all. Eg if I am running xterm or dtterm then it is the terminal itself which defines foreground and background, then the applications can just print text. In bash, I have to strip the colour references from .bashrc to get the effect I want, then whatever client connects gets the correct foreground/background/black/white colours according to their terminal.

**Comment by 237dmitry** [comment]
> but typed text is still yellow

```powershell
Set-PSReadlineOption -color @{
Command = "`e[37m"              
Comment = "`e[37m"
ContinuationPrompt  = "`e[37m"   
Default = "`e[37m"
Emphasis = "`e[37m"
Error = "`e[37m"
InlinePrediction  = "`e[37m"      
Keyword = "`e[37m"
ListPrediction = "`e[37m"
ListPredictionSelected  = "`e[37m"
ListPredictionTooltip  = "`e[37m"
Member = "`e[37m"
Number  = "`e[37m"
Operator  = "`e[37m"
Parameter  = "`e[37m"
Selection =  = "`e[7m"
String  = "`e[37m"
Type =  "`e[37m"
Variable  = "`e[37m" 
}
```

**Comment by rhubarb-geek-nz** [comment]
```
Set-PSReadLineOption: '`e""' is not a valid color value.  It must be a ConsoleColor, ANSI escape sequence, or RGB value with optional leading '#'.
```

I don't want to be setting any colours on the server, the background and foreground colours are defined by the terminal on the client and could be literally anything. The server has no idea what client and terminal theme will be used.

**Comment by 237dmitry** [comment]
I do not understand what you want. If monochrome terminal you have to configure terminal, if monochrome Powershell you have to configure PowerShell.

PS. The error lies in careless copy-paste.

**Comment by rhubarb-geek-nz** [comment]
> PS. The error lies in careless copy-paste.

I was trying to find a null or empty colour setting, but it requires each to be an actual colour. 

A monochrome console app is not one where you have configured the colours to be black and white, it is one where you have told it to not use any colours at all.

**Comment by 237dmitry** [comment]
> It looks like [NO_COLOR](https://no-color.org/) environment variable does what I want

This doesn't work either from the console or from `~/.bashrc`. Maybe this depends on terminal emulator, `Tilix` in my case.

![ss](https://github.com/PowerShell/PowerShell/assets/78153320/aa06f77a-ea8b-491e-bf1c-7520a49613bc)

**Comment by rhubarb-geek-nz** [comment]
> an informal standard was proposed in 2017

Yes, it is a new thing, so not everything supports it yet, but PowerShell itself does! :)

**Comment by 237dmitry** [comment]
> but PowerShell itself does!

It seems not...

![ss](https://github.com/PowerShell/PowerShell/assets/78153320/8ad7b846-e0b1-432a-8d3a-626c461e29fe)

**Comment by rhubarb-geek-nz** [comment]
It works if you set the environment before running `pwsh`

Eg from bash..

```
$ pwsh -c '$PSStyle.OutputRendering'
Host
$ NO_COLOR=true pwsh -c '$PSStyle.OutputRendering'
PlainText
```

**Comment by 237dmitry** [comment]
> Eg from bash..

What is the difference? Look at screenshot once more -- NO_COLOR exported with login...

Ok, new screenshot:

![ss](https://github.com/PowerShell/PowerShell/assets/78153320/e7df0b83-9371-424f-943a-2c5db60a4a09)

**Comment by rhubarb-geek-nz** [comment]
> What is the difference?

Good question, I am running 7.4.1 on both osx and linux and finding that they both respect NO_COLOR.

**Comment by mklement0** [comment]
I can confirm your findings, @rhubarb-geek-nz, on macOS, Linux, and WSL.

Here's the relevant source-code link:

https://github.com/PowerShell/PowerShell/blob/e453c006d39a2fa57a903df2d0ea33ddd0cb43dd/src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs#L83

@237dmitry, the likeliest explanation for the behavior you're seeing is that your profile files negate the effects of the environment variable; try running your commands with `-NoProfile`

**Comment by 237dmitry** [comment]
With `-NoProfile` is `PlainText`

But is it even possible to rely on personal settings in this matter? I think not. Otherwise, it turns out that the option with `NO_COLOR` is not universal. And why it does not work in bash? It ignores `NO_COLOR` environment (deprecated?).

It is easier to set `$PSStyle.OutputRendering = "PlainText"` and configure `PSReadline` (or unload it at all).

**Comment by rhubarb-geek-nz** [comment]
We do want the client to indicate the colour handling, not the server. Consider a shared server with multiple different users logging in with different client systems, some macOS, some windows and some Linux. We want the server to be useable by all of those without having to do lots of custom configuration.

That standard we have had for decades is the TERM variable, where bash seems to enable colourisation when it is set to "xterm-256color" or similar. NO_COLOR is a new addition.

I would ideally like the TERM variable respected as that is automatically relayed by ssh.

**Comment by mklement0** [comment]
@rhubarb-geek-nz, I haven't followed this thread too closely, but
`TERM=xterm-mono pwsh -noprofile -c '$PSStyle.OutputRendering'`
does result in `PlainText`.

**Comment by rhubarb-geek-nz** [comment]
My macOS complains it has no terminal definition for "xterm-mono", and it is not an option in the Terminal Settings, it has xterm, xterm-16color and xterm-256color

If TERM=xterm then OutputRendering is still Host

On a Mac, open a new Terminal Window of type "Man Page" and you will have lovely big yellow window. Run `pwsh` and you can happily type your yellow text on the yellow background.

**Comment by mklement0** [comment]
* PSReadLine is clearly (and unfortunately) a separate story - possibly worth addressing in the latter's repo.

* On my macOS Sonoma system (14.2.1), `TERM=xterm-mono pwsh -noprofile -c '$PSStyle.OutputRendering'`  works as described - not sure about earlier versions.

**Comment by rhubarb-geek-nz** [comment]
On a mac still in zsh

```
 % TERM=xterm-mono
zsh: can't find terminal definition for xterm-mono
```

**Comment by mklement0** [comment]
That is very unfortunate, given that `zsh` is now the recommended user shell, but it appears to be `zsh`'s problem - also worth reporting (to Apple).

**Comment by mklement0** [comment]
To put it differently:

* `bash`, `ksh`, and `dash` happily accept `TERM=xterm-mono pwsh -noprofile -c '$PSStyle.OutputRendering'` - and set the `TERM` to the specified value, _whatever it is_ (including _invalid_ values) - as they should.

* By contrast, `zsh`, seemingly intercepts such child process-scoped environment-variable definitions, at least with respect to `TERM`. Whether or not it is true that `xterm-mono` isn't defined in the macOS terminal database is arguably irrelevant.

**Comment by rhubarb-geek-nz** [comment]
> That is very unfortunate, given that `zsh` is now the recommended user shell, but it appears to be `zsh`'s problem - also worth reporting separately (to Apple).

And 'nano'! I expect it is not in the termcaps database.

**Comment by rhubarb-geek-nz** [comment]
>  Whether or not it is true that `xterm-mono` isn't defined in the macOS terminal database is arguably irrelevant.

Not if then `nano` another programs break on your system because it can't look up the terminal capabilities.

**Comment by mklement0** [comment]
* That `nano` _itself_ also performs such a check is understandable, as it is a program that heavily relies on the capabilities of the terminal - but that is _its_ choice.

* By contrast, a _shell_ interfering with your ability to even _set_ a `TERM` value strike me as inappropriate - notably, `pwsh` (leaving PSReadLine aside) works just fine with an `xterm-mono` value.

> 'pwsh' should look at the termcaps database to see if the terminal is monochrome or not.

`pwsh` - as any utility that cares - should consult the value of environment variable `TERM`, and it does.
It is what tells it what the _current_ terminal session's terminal type is.

**Comment by rhubarb-geek-nz** [comment]
> `pwsh` - as any utility that cares - should consult the value of environment variable `TERM`, and it does. It is what tells it what the _current_ terminal session's terminal type is.

Yes,and then it should use terminfo/termcaps to determine the capabilities of that terminal. So when pwsh is running with TERM=dtterm or TERM=vt100 it can make the correct decisions about how to display information without having to hardcode terminal names. The TERM variable is just a key into the terminfo/termcap database then that tells programs like nano and libraries like curses what control codes to use to drive the terminal.

**Comment by rhubarb-geek-nz** [comment]
On Linux

```
$ for t in xterm-256color xterm-16color xterm xterm-mono vt52 vt100 ansi dtterm; do echo $t $(TERM=$t infocmp | grep 'colors#'); done
xterm-256color colors#0x100, cols#80, it#8, lines#24, pairs#0x10000,
xterm-16color colors#16, cols#80, it#8, lines#24, pairs#0x100,
xterm colors#8, cols#80, it#8, lines#24, pairs#64,
xterm-mono
vt52
vt100
ansi colors#8, cols#80, it#8, lines#24, ncv#3, pairs#64,
dtterm colors#8, cols#80, it#8, lines#24, lm#0, pairs#8,
```

But on a Mac if you use TERM=xterm-mono it breaks nano because it is not in the terminfo.

**Comment by mklement0** [comment]
Let me try to summarize:

* It is up to any given program _whether or not_ to pay attention to the value of the `TERM` environment variable and, if so, _to what extent_.

  * PowerShell does, but only very _selectively_ - it seemingly _only_ uses it to determine whether or not it should modify its `$PSStyle.OutputRendering` value, as described in [the docs](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables#terminal-features).

     * A more _focused_  way to achieve the effect of getting PowerShell to _not_ use coloring (ANSI escape sequences) is to set the `NO_COLOR` environment variable (to _any_ value - all that matter is whether it is _defined or not_).

  * [`PSReadLine`](https://github.com/PowerShell/PSReadLine) - responsible for handling user input in  _interactive_ sessions -  is a separate module that plays by its own rules, and any unexpected/undesirable behavior should be reported in its repo.

* It is inappropriate for a _shell_ to interfere with setting this - or, for that matter,  _any_ - environment variable.

  * That's `zsh`'s problem (which doesn't affect _other_ POSIX-compatible shells, such as `bash`), at least in the implementation that comes with macOS Sonoma (14.x) - and Apple should be told.

**Comment by rhubarb-geek-nz** [comment]
I like the idea of PowerShell core using NO_COLOR for its rendering and then PSReadLine abiding by terminfo. I have raised an [issue in  PSReadLIne](https://github.com/PowerShell/PSReadLine/issues/3918)

**Comment by 237dmitry** [comment]
> Let me try to summarize:

- Without unloading `PSReadline`, works with any configuration:
```
$PSStyle.OutputRendering = "PlainText"
Set-PSReadlineOption -Colors @{
    Command = "`e[37m"              
    Comment = "`e[37m"
    ContinuationPrompt  = "`e[37m"   
    Default = "`e[37m"
    Emphasis = "`e[37m"
    Error = "`e[37m"
    InlinePrediction  = "`e[37m"      
    Keyword = "`e[37m"
    ListPrediction = "`e[37m"
    ListPredictionSelected  = "`e[37m"
    ListPredictionTooltip  = "`e[37m"
    Member = "`e[37m"
    Number  = "`e[37m"
    Operator  = "`e[37m"
    Parameter  = "`e[37m"
    Selection =  "`e[7m"
    String  = "`e[37m"
    Type =  "`e[37m"
    Variable  = "`e[37m" 
}
```

- With unloading `PSReadline` module, works with any configuration:
```
$PSStyle.OutputRendering = "PlainText"
Remove-Module PSReadline
```

- The worst method. Depends on configuration, `-NoProfile` is not a good workaround:
```
$env:NO_COLOR=1
Remove-Module PSReadline
```

**Comment by mklement0** [comment]
P.S.:

That PowerShell _invariably_ loads its profiles _by default_ - even when _unattended_ execution is requested via startup commands (`-File` or `-Command`) - is highly unfortunate, and at odds with POSIX-compatible shells; see:

* #8072
* #3743

**Comment by rhubarb-geek-nz** [comment]
NO_COLOR works with PowerShell/PowerShell. Have raised issue with PowerShell/PSReadLine.

**Comment by rhubarb-geek-nz** [comment]
NO_COLOR works with PowerShell/PowerShell. Have raised issue with PowerShell/PSReadLine.

**Comment by 237dmitry** [comment]
> NO_COLOR works with PowerShell/PowerShell.

Maybe on MacOs, on Linux it requires additional configuration.

**Comment by rhubarb-geek-nz** [comment]
> Maybe on MacOs, on Linux it requires additional configuration.

It works on all my Linux installations which come from deb, rpm or apk. ( It works on my machine™ )

On Ubuntu 22.04

```
$ pwsh -c '$PSStyle.OutputRendering'
Host
$ NO_COLOR=true pwsh -c '$PSStyle.OutputRendering'
PlainText
```

On RedHat 9

```
$ pwsh -c '$PSStyle.OutputRendering'
Host
$ NO_COLOR=true pwsh -c '$PSStyle.OutputRendering'
PlainText
```

On Alpine 3.19

```
$ pwsh -c '$PSStyle.OutputRendering'
Host
$ NO_COLOR=true pwsh -c '$PSStyle.OutputRendering'
PlainText
```

**Comment by 237dmitry** [comment]
> ... and no custom user profile.

Honestly, it seems very strange.

**Comment by mklement0** [comment]
Let me try another, more comprehensive summary:

* On the _output_ side, to get _PowerShell itself_ not to emit ANSI/VT escape sequences  (not to use colors) in for-display output, you have three options:

  * In-session:
    * `$PSStyle.OutputRendering = 'PlainText'` (may be placed in `$PROFILE`); this is the only way to guarantee that the effects are limited to PowerShell.

  * Before startup, via environment variables, which make PowerShell _automatically_ put `$PSStyle.OutputRendering = 'PlainText'` into effect:
 
    * By defining `NO_COLOR`, e.g. with a value of `1`, a _convention_ proposed by [no-color.org](https://no-color.org/).

      * Actually, the specific value doesn't matter, as long as is not the empty string (per no-color.org); however, PowerShell [accepts the empty string too](https://github.com/PowerShell/PowerShell/blob/e453c006d39a2fa57a903df2d0ea33ddd0cb43dd/src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterface.cs#L83-L86).
      * Other programs may or may not honor this convention, though the number that do will likely grow over time. 
      * Unlike the `TERM` variable (see below), this variable represents an abstract setting that simply communicates the user's preference for programs not to use colors.
   
     * Alternatively, by setting `TERM` to either `xterm-mono` (Linux and Windows) or `xtermm` (macOS and Windows) (see [about_ANSI_Terminals](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_ANSI_Terminals)), but this comes with caveats:
       * On Windows, the `TERM` environment variable has no special meaning, but is honored by PowerShell by convention, for cross-platform consistency.     
       * On Unix-like platforms, setting the [POSIX-defined `TERM` environment variable](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html#tag_08) affects all programs that rely on the value to reflect an actual terminal name (e.g. `xterm-color256`) that can be looked up in the host platform's terminal database.
         * Terminal-focused programs such as the `nano` editor refuse to run if the `TERM` value has no terminal-database entry.
           *  On macOS (but not Linux) this problem happens with value `xterm-mono` (which is fine on Linux), so `xtermm` must be used instead.
           * Curiously, `zsh`, the default user shell on macOS, actively _prevents_ setting `TERM` to an unknown value - unlike other POSIX-compatible shells such as `bash`.
         * To test support for a given terminal name on Unix-like platforms, use `infocmp -t <name>`
         * Note that PowerShell itself is _not_ concerned with specific terminal capabilities - it merely uses the `TERM` value to govern its output-coloring behavior.

       * A value of `dumb` is also accepted, but best avoided:
         * This terminal name signals the absence of VT support (ANSI/VT escape-sequence support) in the current terminal, which, on Unix-like platforms, makes `PSReadLine` _malfunction_ and also prevents terminal-focused programs such as `nano` from running.
         * While this value too _effectively_ disables output coloring, it does so _not_ by setting `$PSStyle.OutputRendering = 'PlainText'` (it'll be at its default, `Host`), but by detecting the implied absence of VT support and making the read-only `$host.ui.SupportsVirtualTerminal` property report `$false`.
In other words: the value of `$PSStyle.OutputRendering` is then _irrelevant_.

     * It follows from the above that the environment-variable based approach:
       * affects other programs too, on Unix-like platforms
       * acts as a _preset_ that _can be overridden in-session_ by modifying `$PSStyle.OutputRendering` - _except_ if `TERM` is set to `dumb`.

   * Related problems and limitations, as of PowerShell 7.4.1:
     * #20170
     * #20171
     * #17509

* On the interactive _input_ side, user input is handled by the [`PSReadLine`](https://github.com/PowerShell/PSReadLine) module, not PowerShell, and `PSReadLine` currently (v2.3.4) does _not_ honor any of these settings, and _always_ uses colors.
  * https://github.com/PowerShell/PSReadLine/issues/3918 aims to change that.

**Comment by rhubarb-geek-nz** [comment]
zsh does not prevent TERM being set, but does warn.

```
% TERM=xterm-mono
zsh: can't find terminal definition for xterm-mono
% echo $TERM
xterm-mono
```

While Windows does not currently use `TERM`, this variable is faithfully relayed by the Microsoft provided `ssh`. If `TERM` is not set then it appears to use `xterm-256color`.

The Windows Win32 Console API does work over SSH using terminal escape codes. I don't know how it determines the client terminal capabilities or just assumes `ansi`.

With Windows Servers being accessed and managed by SSH, the variable may become more significant.

Let us not forget [Default CLI Editor - Feature Exploration!](https://github.com/microsoft/terminal/discussions/16440)

**Comment by mklement0** [comment]
Thanks - didn't realize that `zsh` only _warns_ - I've updated my previous comments accordingly.

**Comment by rhubarb-geek-nz** [comment]
> P.S.: It is part of my previous comment, but to highlight it explicitly: the macOS equivalent of `TERM=xterm-mono` is `TERM=xtermm`

Thanks, with a quick test, `xtermm` is recognized by Debian and CM Mariner as monochrome.

**Comment by mklement0** [comment]
Unfortunately, not all Linux distros recognize it, however (tested via `infocmp -t xtermm`): 
* Ubuntu 22.04.03 doesn't (at least via WSL)
* [CBL-Mariner](https://github.com/microsoft/CBL-Mariner), at least with a bare-bones Docker image, doesn't.

**Comment by rhubarb-geek-nz** [comment]
On CBL-Mariner you need ncurses-term

```
$ rpm -ql ncurses-term | grep xtermm
/usr/share/terminfo/x/xtermm
```

Likewise with Ubuntu 22.04

```
$ dpkg-query -S $(find /usr -name xtermm -type f)
ncurses-term: /usr/share/terminfo/x/xtermm
```

**Comment by mklement0** [comment]
Thanks; the **pragmatic conclusion** is:

* On the output side: While the lack of standardized terminal names across Unix-like platforms _out of the box_ is unfortunate, choosing to set the `NO_COLOR` environment variable instead bypasses the problem and predictably works on all supported platforms.
  * An in-session, PowerShell-only solution is to set `$PSStyle.OutputRendering = 'PlainText'` from a `$PROFILE` file - though this can be bypassed via the `-NoProfile` CLI switch.

* On the interactive input side: It is to be hoped for that a future `PSReadLine` version will honor `NO_COLOR` / `TERM` and possibly `$PSStyle.OutputRendering` too - [details to be discussed](https://github.com/PowerShell/PSReadLine/issues/3918).
